hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/util/BoundPriorityQueue.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.tpc.util;
+
+import java.util.PriorityQueue;
+
+/**
+ * A bound version of the {@link PriorityQueue}. Instead of having the backing
+ * array grow, the array has a fixed size. So no growing and control on the maximum
+ * number of items in this priority queue.
+ *
+ * @param <E>
+ */
+public class BoundPriorityQueue<E> extends PriorityQueue<E> {
+
+    private final int capacity;
+
+    /**
+     * Creates a BoundPriorityQueue with the given capacity.
+     *
+     * @param capacity the capacity.
+     * @throws IllegalArgumentException when capacity smaller than 0.
+     */
+    public BoundPriorityQueue(int capacity) {
+        super(capacity);
+        this.capacity = capacity;
+    }
+
+    @Override
+    public boolean offer(E e) {
```

Comment:
Should we also guard the add and addAll methods? From the Javadocs of these we can see that add uses offer underneath (mentioned in the superclass https://docs.oracle.com/javase/8/docs/api/java/util/AbstractQueue.html, but not here https://docs.oracle.com/javase/8/docs/api/java/util/PriorityQueue.html) but the Javadoc for addAll is kinda vague

Reply:
I see that when I looked at the implementation, but can we be sure about that from the contract of these APIs? As I said, it is written in the Javadoc of https://docs.oracle.com/javase/8/docs/api/java/util/AbstractQueue.html that add is implemented on top of offer, so we should be fine there, but I am not sure about addAll. It just says `This implementation iterates over the specified collection, and adds each element returned by the iterator to this queue, in turn.` There is no mention of using the `add` or `offer` methods to add the item.

Reply:
```
  public boolean addAll(Collection<? extends E> c) {
        if (c == null)
            throw new NullPointerException();
        if (c == this)
            throw new IllegalArgumentException();
        boolean modified = false;
        for (E e : c)
            if (add(e))<----
                modified = true;
        return modified;
    }
 ```

Reply:
I saw that as well, but isn't it just an implementation detail? Anyways, I think any reasonable implementation would re-use add here, even though it is not clearly written in the Javadoc. So, I am resolving this